src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -138,6 +144,58 @@ public void End(ProduceEndType endType)
             }
         }
 
+        public MemoryPoolIterator2 ProducingStart()
+        {
+            lock (_returnLock)
+            {
+                Debug.Assert(!_isProducing);
+                _isProducing = true;
+
+                if (_tail == null)
+                {
+                    throw new IOException("The socket has been closed.");
+                }
+
+                return new MemoryPoolIterator2(_tail, _tail.End);
```

Comment:
The lock here is not to protect against overlapping writes from user code.

This is to prevent NullRefExceptions when the socket is closed on the uv thread (which calls ReturnAllBlocks and nulls out _tail) during the call to ProducingStart.